torch/_higher_order_ops/cond.py

```diff
@@ -142,12 +146,111 @@ def _validate_input(pred, true_fn, false_fn, operands):
                     pred, true_fn, false_fn, operands
                 )
 
-
 """
 We're going to define a `cond_op` operation.
 In order to do this, we need implementations for each of the dispatch keys.
 """
 cond_op = HigherOrderOperator("cond")
+cond_op.__module__ = "torch.ops.higher_order"
+
+def create_fw_bw_graph(true_fn, false_fn, *operands):
```

Comment:
I created such a low-level function as suggested and it makes sense to me. I also integrated this low-level function into the [while loop](https://github.com/pytorch/pytorch/blob/8008d5dbfc9588aa27aa1a5380ae5c51b3b57900/torch/_higher_order_ops/while_loop.py#L96) in this [PR](https://github.com/pytorch/pytorch/pull/124573).

However, there is one problem with that I realized: The `create_fw_bw_graph` function of the `cond`, or any other higher_order_op is invoked twice. Once in the forward pass, which expected. However, it is invoked a second time during the backward pass, which causes some trouble. In fact, I had to add a "detection mechanism" for that case and avoid creating a `joint_fn` of an already created `joint_fn`. If this second invocation can be removed, it would simplify things quite a bit.

Reply:
@ydwu4 What do you think about this lower-level function in `utils.py`?

As I mentioned, I still see this problem that the `create_fw_bw_graph_branches` function is invoked in the forward and in the backward path. For `cond` it might not be too much of a problem, but for other operations, such as `while_loop`, it may be problematic. 

This is because in the case of the while_loop for example, the cond and the body function in the backward path have slightly different signatures compared to the forward path. 
```
def cond_fn(x):
    return x.sum() < 10
```
vs.
```
def bw_cond_fn(*args):
    return len(args[-1]) > 1
```
where `args` is a tuple of `(grads, fw_outputs)`



For the first invokation of `create_fw_bw_graph_branches` (in the forward path) this is fine, but it causes problems when this function is called again in the backward path. One has to explicitly detect that one is in the backward path and avoid calling the lower-level function, but rather just return dummy graphs, as they are not used anyways. [Here](https://github.com/pytorch/pytorch/blob/b7aefd85743e51b708fcb6b61e43fb330f7ee25d/torch/_higher_order_ops/while_loop.py#L97-L105) is an example of what I mean.
Do you have any thoughts on this?

Reply:
We can make fw_outputs an argument instead of computing it inside of the interface. https://github.com/pytorch/pytorch/pull/126911/files#diff-94838db6991b14cfeec36d4eb0e0ff6fadc08624a7c3ac967820f7f7ae267e70R294 

BTW, let's pause while_loop and bit and I'll need some time to think this through.

Reply:
Makes sense to me, I have updated the [low-level function](https://github.com/pytorch/pytorch/blob/630549a7e12c32b0f0f9a366c53e247b9a32a230/torch/_higher_order_ops/utils.py#L266) accordingly.

I have already paused the development on the while_loop as you suggested. However, I just tried to have it in mind for the `create_fw_bw_graph` function. In general, I am quite eager to have the while_loop feature ready, so whenever there is something that I can help, please do let me know.